File: framework/languages/dotnet/samples/SupportChat/Server/Support/Infrastructure/ZLink/Spots/Handlers/SetAgentAvailableHandler.cs
using SupportChat.Server.Support.Infrastructure.ZLink.Actors;
using SupportChat.Server.Support.Application.ConversationAssignment;
using SupportChat.Server.Configuration;
using SupportChat.Shared.Contracts;
using Zlink.Framework.Contracts.Actors;
using Zlink.Framework.Contracts.Spots;

namespace SupportChat.Server.Support.Infrastructure.ZLink.Spots.Handlers;

internal sealed class SetAgentAvailableHandler(
    AgentAvailabilityDirectory availability,
    IZLinkActorManager actorManager,
    SupportActorDirectory actors)
    : IZLinkEntrySpotActorRequestHandler<SupportEntrySpot, SupportUserActor, SetAgentAvailableReq, SetAgentAvailableRes>
{
    public async ValueTask<SetAgentAvailableRes> HandleAsync(
        SupportEntrySpot entrySpot,
        SupportUserActor actor,
        ZLinkSpotActorRequestContext context,
        SetAgentAvailableReq message,
        CancellationToken cancellationToken)
    {
        _ = entrySpot;
        _ = context;
        cancellationToken.ThrowIfCancellationRequested();
        if (!string.Equals(actor.Role, SupportChatRoles.Agent, StringComparison.Ordinal))
        {
            throw new InvalidOperationException("Only agent actors can set availability.");
        }

        var actorRef = await actorManager.FindAsync(actor.ActorId, cancellationToken)
            ?? throw new InvalidOperationException($"Support actor ref is not available. actor={actor.ActorId}");
        actors.AddOrUpdate(actor, actorRef);
        availability.SetAvailable(actor.ActorId, actor.DisplayName, message.IsAvailable);
        return new SetAgentAvailableRes(message.IsAvailable);
    }
}

File: framework/languages/node/samples/SupportChat.Ts/Server/Support/Infrastructure/ZLink/Spots/Handlers/set-agent-available-handler.ts
import { zlinkEntrySpotActorRequestHandler } from '@zlink-systems/nestjs';
import { SupportEntrySpot } from '../support-entry-spot';
import { SupportUserActor } from '../../Actors/support-user-actor';
import { PacketNames } from '../../../../../../Shared/Contracts/messages';
import type {
  ZLinkEntrySpotActorRequestHandler,
  ZLinkSpotActorRequestContext
} from '@zlink-systems/framework';
import type { SupportEntrySpot as SupportEntrySpotType } from '../support-entry-spot';
import type { SupportUserActor as SupportUserActorType } from '../../Actors/support-user-actor';
import type {
  SetAgentAvailableReq,
  SetAgentAvailableRes
} from '../../../../../../Shared/Contracts/messages';

@zlinkEntrySpotActorRequestHandler({
  actor: () => SupportUserActor,
  entrySpot: () => SupportEntrySpot,
  packetName: PacketNames.setAgentAvailableReq
})
class SetAgentAvailableHandler
  implements ZLinkEntrySpotActorRequestHandler<SupportEntrySpotType, SupportUserActorType, SetAgentAvailableReq, SetAgentAvailableRes> {
  async handle(
    entrySpot: SupportEntrySpotType,
    actor: SupportUserActorType,
    context: ZLinkSpotActorRequestContext,
    request: SetAgentAvailableReq
  ): Promise<SetAgentAvailableRes> {
    void context;
    return entrySpot.setAgentAvailable({
      ...request,
      actorId: actor.actorId,
      displayName: actor.displayName,
      role: actor.role
    });
  }
}

export { SetAgentAvailableHandler };
